File: src/lib/registry.ts
/**
 * Live skill package registry — fetches from GitHub, caches locally, falls back to hardcoded.
 *
 * Sources:
 *   1. OpenAEC Foundation org — individual repos matching *Skill-Package* or *Claude-Skill*
 *   2. Anthropic official — anthropics/skills repo with marketplace.json manifest
 */

import { packages as hardcodedPackages, type SkillPackage } from "./packages";

export type Publisher = "openaec" | "anthropic";

export interface RegistryPackage extends SkillPackage {
  publisher: Publisher;
  repoUrl: string;
  updatedAt: string;
  skillsPath: string; // relative path to skills dir in repo
}

interface GitHubRepo {
  name: string;
  full_name: string;
  description: string | null;
  html_url: string;
  updated_at: string;
  topics: string[];
}

interface AnthropicMarketplace {
  name: string;
  plugins: {
    name: string;
    description: string;
    skills: string[];
  }[];
}

// Cache duration: 10 minutes
const CACHE_KEY = "openaec-registry-cache";
const CACHE_TTL = 10 * 60 * 1000;

interface CacheEntry {
  packages: RegistryPackage[];
  timestamp: number;
}

function getCached(): RegistryPackage[] | null {
  try {
    const raw = localStorage.getItem(CACHE_KEY);
    if (!raw) return null;
    const entry: CacheEntry = JSON.parse(raw);
    if (Date.now() - entry.timestamp > CACHE_TTL) return null;
    return entry.packages;
  } catch {
    return null;
  }
}

function setCache(packages: RegistryPackage[]) {
  try {
    const entry: CacheEntry = { packages, timestamp: Date.now() };
    localStorage.setItem(CACHE_KEY, JSON.stringify(entry));
  } catch {
    // localStorage not available (e.g. in Tauri webview with restrictions)
  }
}

/** Parse skill count from GitHub description like "27 deterministic Claude AI skills for..." */
function parseSkillCount(description: string | null): number {
  if (!description) return 0;
  const match = description.match(/^(\d+)\s+(deterministic\s+)?/i);
  return match ? parseInt(match[1], 10) : 0;
}

/** Infer category from repo topics or name */
function inferCategory(repo: GitHubRepo): SkillPackage["category"] {
  const name = repo.name.toLowerCase();
  const topics = repo.topics.map((t) => t.toLowerCase());
  const all = [name, ...topics].join(" ");

  if (all.includes("blender") || all.includes("bonsai") || all.includes("ifc") || all.includes("bim") || all.includes("speckle") || all.includes("qgis") || all.includes("thatopen") || all.includes("three")) return "aec-bim";
  if (all.includes("erpnext") || all.includes("nextcloud") || all.includes("n8n")) return "erp-business";
  if (all.includes("docker") || all.includes("draw")) return "devops";
  if (all.includes("cross-tech")) return "cross-tech";
  return "web-dev";
}

/** Infer package ID from repo name */
function repoToId(repoName: string): string {
  return repoName
    .replace(/-Claude-Skill-Package$/i, "")
    .replace(/_Anthropic_Claude_Development_Skill_Package$/i, "")
    .replace(/-ifcOpenshell-Sverchok$/i, "")
    .toLowerCase()
    .replace(/_/g, "-");
}

/** Infer status from skill count and description */
function inferStatus(skillCount: number, description: string | null): SkillPackage["status"] {
  if (skillCount === 0 || !description) return "planned";
  if (skillCount < 10) return "development";
  return "published";
}

/** Fetch OpenAEC Foundation skill packages from GitHub API */
async function fetchOpenAECPackages(): Promise<RegistryPackage[]> {
  const response = await fetch(
    "https://api.github.com/orgs/OpenAEC-Foundation/repos?per_page=100&type=public",
    { headers: { Accept: "application/vnd.github.v3+json" } }
  );
  if (!response.ok) throw new Error(`GitHub API: ${response.status}`);

  const repos: GitHubRepo[] = await response.json();
  const skillRepos = repos.filter(
    (r) => r.name.includes("Skill-Package") || r.name.includes("Claude-Skill")
  );

  // Exclude the template repo
  const filtered = skillRepos.filter(
    (r) => !r.name.includes("Workflow-Template")
  );

  return filtered.map((repo) => {
    const skillCount = parseSkillCount(repo.description);
    const id = repoToId(repo.name);
    return {
      id,
      name: repo.name
        .replace(/-Claude-Skill-Package$/i, "")
        .replace(/_Anthropic_Claude_Development_Skill_Package$/i, "")
        .replace(/-/g, " ")
        .replace(/_/g, " "),
      description: repo.description || "",
      category: inferCategory(repo),
      skillCount,
      repo: repo.full_name,
      repoUrl: repo.html_url,
      status: inferStatus(skillCount, repo.description),
      tags: repo.topics.length > 0 ? repo.topics : [id],
      publisher: "openaec" as Publisher,
      updatedAt: repo.updated_at,
      skillsPath: "skills/source",
    };
  });
}

/** Fetch Anthropic official skills from marketplace.json */
async function fetchAnthropicPackages(): Promise<RegistryPackage[]> {
  const response = await fetch(
    "https://api.github.com/repos/anthropics/skills/contents/.claude-plugin/marketplace.json",
    { headers: { Accept: "application/vnd.github.v3+json" } }
  );
  if (!response.ok) throw new Error(`GitHub API: ${response.status}`);

  const file = await response.json();
  const content = atob(file.content);
  const manifest: AnthropicMarketplace = JSON.parse(content);

  // Also get repo metadata for updated_at
  const repoResponse = await fetch(
    "https://api.github.com/repos/anthropics/skills",
    { headers: { Accept: "application/vnd.github.v3+json" } }
  );
  const repoMeta = repoResponse.ok ? await repoResponse.json() : { updated_at: new Date().toISOString() };

  const packages: RegistryPackage[] = [];

  for (const plugin of manifest.plugins) {
    packages.push({
      id: `anthropic-${plugin.name}`,
      name: `${plugin.name} (Anthropic)`,
      description: plugin.description,
      category: "web-dev" as const,
      skillCount: plugin.skills.length,
      repo: "anthropics/skills",
      repoUrl: "https://github.com/anthropics/skills",
      status: "published" as const,
      tags: plugin.skills.map((s) => s.replace("./skills/", "")),
      publisher: "anthropic" as Publisher,
      updatedAt: repoMeta.updated_at,
      skillsPath: "skills",
    });
  }

  return packages;
}

/**
 * Fetch all packages from all sources.
 * Returns cached data if available, fetches fresh data in background.
 */
export async function fetchRegistry(): Promise<RegistryPackage[]> {
  // Return cache if fresh
  const cached = getCached();
  if (cached) return cached;

  try {
    const [openaec, anthropic] = await Promise.allSettled([
      fetchOpenAECPackages(),
      fetchAnthropicPackages(),
    ]);

    const packages: RegistryPackage[] = [
      ...(openaec.status === "fulfilled" ? openaec.value : []),
      ...(anthropic.status === "fulfilled" ? anthropic.value : []),
    ];

    if (packages.length > 0) {
      setCache(packages);
      return packages;
    }
  } catch {
    // Network error — fall through to hardcoded
  }

  // Fallback: convert hardcoded packages to RegistryPackage format
  return hardcodedPackages.map((p) => ({
    ...p,
    publisher: "openaec" as Publisher,
    repoUrl: p.repo ? `https://github.com/${p.repo}` : "",
    updatedAt: "",
    skillsPath: "skills/source",
  }));
}

/**
 * Get hardcoded packages as RegistryPackage (instant, for initial render).
 */
export function getHardcodedRegistry(): RegistryPackage[] {
  return hardcodedPackages.map((p) => ({
    ...p,
    publisher: "openaec" as Publisher,
    repoUrl: p.repo ? `https://github.com/${p.repo}` : "",
    updatedAt: "",
    skillsPath: "skills/source",
  }));
}

/**
 * Fetch registry via Rust backend (preferred — no CORS, disk cache, higher rate limits)
 */
export async function fetchRegistryBackend(forceRefresh = false): Promise<RegistryPackage[]> {
  try {
    const { invoke } = await import("@tauri-apps/api/core");
    const packages = await invoke<any[]>("fetch_registry", { forceRefresh });
    return packages.map(p => ({
      id: p.id,
      name: p.name,
      description: p.description,
      category: p.category as SkillPackage["category"],
      skillCount: p.skill_count,
      repo: p.repo,
      repoUrl: p.repo_url,
      status: p.status as SkillPackage["status"],
      tags: p.tags,
      publisher: p.publisher as Publisher,
      updatedAt: p.updated_at,
      skillsPath: p.skills_path,
    }));
  } catch {
    // Fallback to frontend fetch
    return fetchRegistry();
  }
}

/**
 * Get cached registry from Rust backend (instant, for offline/startup)
 */
export async function getCachedRegistryBackend(): Promise<RegistryPackage[]> {
  try {
    const { invoke } = await import("@tauri-apps/api/core");
    const packages = await invoke<any[]>("get_cached_registry");
    return packages.map(p => ({
      id: p.id,
      name: p.name,
      description: p.description,
      category: p.category as SkillPackage["category"],
      skillCount: p.skill_count,
      repo: p.repo,
      repoUrl: p.repo_url,
      status: p.status as SkillPackage["status"],
      tags: p.tags,
      publisher: p.publisher as Publisher,
      updatedAt: p.updated_at,
      skillsPath: p.skills_path,
    }));
  } catch {
    return getHardcodedRegistry();
  }
}

File: src-tauri/src/registry.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegistryPackage {
    pub id: String,
    pub name: String,
    pub description: String,
    pub category: String,
    pub skill_count: u32,
    pub repo: String,
    pub repo_url: String,
    pub status: String,
    pub tags: Vec<String>,
    pub publisher: String,
    pub updated_at: String,
    pub skills_path: String,
    /// Individual skill names found in the repo
    #[serde(default)]
    pub skill_names: Vec<String>,
}

#[derive(Debug, Deserialize)]
struct GitHubRepo {
    name: String,
    full_name: String,
    description: Option<String>,
    html_url: String,
    updated_at: String,
    topics: Vec<String>,
}

#[derive(Debug, Deserialize)]
struct MarketplacePlugin {
    name: String,
    description: String,
    skills: Vec<String>,
}

#[derive(Debug, Deserialize)]
struct MarketplaceManifest {
    plugins: Vec<MarketplacePlugin>,
}

/// GitHub Git Tree API response (efficient: 1 call gets entire tree)
#[derive(Debug, Deserialize)]
struct GitTreeResponse {
    tree: Vec<GitTreeEntry>,
    truncated: bool,
}

#[derive(Debug, Deserialize)]
struct GitTreeEntry {
    path: String,
    #[serde(rename = "type")]
    entry_type: String,
}

/// GitHub Contents API response entry
#[derive(Debug, Deserialize)]
struct GitHubContent {
    name: String,
    #[serde(rename = "type")]
    content_type: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct CacheEntry {
    packages: Vec<RegistryPackage>,
    timestamp: u64,
}

fn cache_dir() -> PathBuf {
    dirs::config_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join("openaec-composer")
        .join("cache")
}

fn cache_file() -> PathBuf {
    cache_dir().join("registry.json")
}

fn get_cached(max_age_secs: u64) -> Option<Vec<RegistryPackage>> {
    let path = cache_file();
    if !path.exists() {
        return None;
    }
    let content = fs::read_to_string(&path).ok()?;
    let entry: CacheEntry = serde_json::from_str(&content).ok()?;
    let now = SystemTime::now().duration_since(UNIX_EPOCH).ok()?.as_secs();
    if now - entry.timestamp > max_age_secs {
        return None;
    }
    Some(entry.packages)
}

fn set_cache(packages: &[RegistryPackage]) {
    let _ = fs::create_dir_all(cache_dir());
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();
    let entry = CacheEntry {
        packages: packages.to_vec(),
        timestamp: now,
    };
    if let Ok(json) = serde_json::to_string_pretty(&entry) {
        let _ = fs::write(cache_file(), json);
    }
}

fn parse_skill_count(description: &Option<String>) -> u32 {
    description
        .as_ref()
        .and_then(|d| d.split_whitespace().next()?.parse::<u32>().ok())
        .unwrap_or(0)
}

fn infer_category(name: &str, topics: &[String]) -> String {
    let all = format!("{} {}", name.to_lowercase(), topics.join(" ").to_lowercase());
    if all.contains("blender") || all.contains("bonsai") || all.contains("ifc")
        || all.contains("bim") || all.contains("speckle") || all.contains("qgis")
        || all.contains("thatopen") || all.contains("three")
    {
        "aec-bim".to_string()
    } else if all.contains("erpnext") || all.contains("frappe") || all.contains("nextcloud") || all.contains("n8n") {
        "erp-business".to_string()
    } else if all.contains("docker") || all.contains("draw") {
        "devops".to_string()
    } else if all.contains("cross-tech") || all.contains("open-pdf-studio") {
        "cross-tech".to_string()
    } else {
        "web-dev".to_string()
    }
}

fn infer_status(skill_count: u32, has_topics: bool) -> String {
    if skill_count >= 10 {
        "published".to_string()
    } else if skill_count > 0 || has_topics {
        "development".to_string()
    } else {
        "planned".to_string()
    }
}

fn repo_to_id(name: &str) -> String {
    name.replace("-Claude-Skill-Package", "")
        .replace("_Claude_Skill_Package", "")
        .replace("_Anthropic_Claude_Development_Skill_Package", "")
        .replace("-ifcOpenshell-Sverchok", "")
        .to_lowercase()
        .replace('_', "-")
}

fn clean_display_name(name: &str) -> String {
    name.replace("-Claude-Skill-Package", "")
        .replace("_Claude_Skill_Package", "")
        .replace("_Anthropic_Claude_Development_Skill_Package", "")
        .replace('-', " ")
        .replace('_', " ")
}

/// Count SKILL.md files in a repo using the Git Tree API (1 API call for entire repo tree).
/// Returns (skill_count, skill_names).
async fn count_real_skills(
    client: &reqwest::Client,
    full_name: &str,
) -> (u32, Vec<String>) {
    // Use the Trees API with recursive=1 to get the full file tree in one call
    let url = format!(
        "https://api.github.com/repos/{}/git/trees/main?recursive=1",
        full_name
    );
    let response = match client
        .get(&url)
        .header("Accept", "application/vnd.github.v3+json")
        .header("User-Agent", "OpenAEC-Workspace-Composer")
        .send()
        .await
    {
        Ok(r) if r.status().is_success() => r,
        _ => return (0, vec![]),
    };

    let tree: GitTreeResponse = match response.json().await {
        Ok(t) => t,
        Err(_) => return (0, vec![]),
    };

    // Find all SKILL.md files → each one = one skill
    let skill_entries: Vec<&GitTreeEntry> = tree
        .tree
        .iter()
        .filter(|e| e.entry_type == "blob" && e.path.ends_with("/SKILL.md"))
        .collect();

    let skill_names: Vec<String> = skill_entries
        .iter()
        .filter_map(|e| {
            // Extract skill name from path like "skills/source/group/skill-name/SKILL.md"
            let parts: Vec<&str> = e.path.split('/').collect();
            if parts.len() >= 2 {
                Some(parts[parts.len() - 2].to_string())
            } else {
                None
            }
        })
        .collect();

    (skill_entries.len() as u32, skill_names)
}

async fn fetch_openaec_packages(
    client: &reqwest::Client,
) -> Result<Vec<RegistryPackage>, String> {
    let response = client
        .get("https://api.github.com/orgs/OpenAEC-Foundation/repos?per_page=100&type=public")
        .header("Accept", "application/vnd.github.v3+json")
        .header("User-Agent", "OpenAEC-Workspace-Composer")
        .send()
        .await
        .map_err(|e| format!("GitHub API error: {}", e))?;

    if !response.status().is_success() {
        return Err(format!("GitHub API returned {}", response.status()));
    }

    let repos: Vec<GitHubRepo> = response.json().await
        .map_err(|e| format!("Failed to parse response: {}", e))?;

    let skill_repos: Vec<GitHubRepo> = repos
        .into_iter()
        .filter(|r| r.name.contains("Skill-Package") || r.name.contains("Claude-Skill") || r.name.contains("Skill_Package"))
        .filter(|r| !r.name.contains("Workflow-Template"))
        .collect();

    // Fetch real skill counts in parallel (up to 20 concurrent)
    let mut packages = Vec::new();
    for repo in &skill_repos {
        let description_count = parse_skill_count(&repo.description);
        let (real_count, skill_names) = count_real_skills(client, &repo.full_name).await;

        // Use the higher of description count vs real count
        let skill_count = if real_count > 0 { real_count } else { description_count };
        let id = repo_to_id(&repo.name);

        packages.push(RegistryPackage {
            id,
            name: clean_display_name(&repo.name),
            description: repo.description.clone().unwrap_or_default(),
            category: infer_category(&repo.name, &repo.topics),
            skill_count,
            repo: repo.full_name.clone(),
            repo_url: repo.html_url.clone(),
            status: infer_status(skill_count, !repo.topics.is_empty()),
            tags: if repo.topics.is_empty() {
                vec![repo_to_id(&repo.name)]
            } else {
                repo.topics.clone()
            },
            publisher: "openaec".to_string(),
            updated_at: repo.updated_at.clone(),
            skills_path: "skills/source".to_string(),
            skill_names,
        });
    }

    Ok(packages)
}

async fn fetch_anthropic_packages(
    client: &reqwest::Client,
) -> Result<Vec<RegistryPackage>, String> {
    // 1. Fetch marketplace.json for plugin bundles
    let response = client
        .get("https://api.github.com/repos/anthropics/skills/contents/.claude-plugin/marketplace.json")
        .header("Accept", "application/vnd.github.v3+json")
        .header("User-Agent", "OpenAEC-Workspace-Composer")
        .send()
        .await
        .map_err(|e| format!("GitHub API error: {}", e))?;

    if !response.status().is_success() {
        return Err(format!("GitHub API returned {}", response.status()));
    }

    #[derive(Deserialize)]
    struct GitHubFile { content: String }
    let file: GitHubFile = response.json().await
        .map_err(|e| format!("Failed to parse: {}", e))?;

    let clean_content = file.content.replace('\n', "");
    use base64::Engine;
    let decoded = base64::engine::general_purpose::STANDARD
        .decode(&clean_content)
        .map_err(|e| format!("Base64 decode error: {}", e))?;

    let manifest: MarketplaceManifest = serde_json::from_slice(&decoded)
        .map_err(|e| format!("JSON parse error: {}", e))?;

    // 2. Also list individual skills from skills/ directory
    let skills_response = client
        .get("https://api.github.com/repos/anthropics/skills/contents/skills")
        .header("Accept", "application/vnd.github.v3+json")
        .header("User-Agent", "OpenAEC-Workspace-Composer")
        .send()
        .await;

    let individual_skills: Vec<String> = if let Ok(resp) = skills_response {
        if resp.status().is_success() {
            let contents: Vec<GitHubContent> = resp.json().await.unwrap_or_default();
            contents.iter()
                .filter(|c| c.content_type == "dir")
                .map(|c| c.name.clone())
                .collect()
        } else {
            vec![]
        }
    } else {
        vec![]
    };

    let mut packages = Vec::new();

    // Add plugin bundles
    for plugin in &manifest.plugins {
        let skill_names: Vec<String> = plugin.skills.iter()
            .map(|s| s.replace("./skills/", ""))
            .collect();

        packages.push(RegistryPackage {
            id: format!("anthropic-{}", plugin.name),
            name: format!("{} (Anthropic)", plugin.name),
            description: plugin.description.clone(),
            category: if plugin.name.contains("document") { "devops".to_string() } else { "web-dev".to_string() },
            skill_count: plugin.skills.len() as u32,
            repo: "anthropics/skills".to_string(),
            repo_url: "https://github.com/anthropics/skills".to_string(),
            status: "published".to_string(),
            tags: skill_names.clone(),
            publisher: "anthropic".to_string(),
            updated_at: String::new(),
            skills_path: "skills".to_string(),
            skill_names,
        });
    }

    // Add individual skills that aren't in any bundle
    let bundled: Vec<String> = manifest.plugins.iter()
        .flat_map(|p| p.skills.iter().map(|s| s.replace("./skills/", "")))
        .collect();

    for skill in &individual_skills {
        if !bundled.contains(skill) {
            packages.push(RegistryPackage {
                id: format!("anthropic-{}", skill),
                name: format!("{} (Anthropic)", skill),
                description: format!("Anthropic official skill: {}", skill),
                category: "web-dev".to_string(),
                skill_count: 1,
                repo: "anthropics/skills".to_string(),
                repo_url: format!("https://github.com/anthropics/skills/tree/main/skills/{}", skill),
                status: "published".to_string(),
                tags: vec![skill.clone(), "anthropic".to_string()],
                publisher: "anthropic".to_string(),
                updated_at: String::new(),
                skills_path: format!("skills/{}", skill),
                skill_names: vec![skill.clone()],
            });
        }
    }

    Ok(packages)
}

#[tauri::command]
pub async fn fetch_registry(force_refresh: bool) -> Result<Vec<RegistryPackage>, String> {
    if !force_refresh {
        if let Some(cached) = get_cached(600) {
            return Ok(cached);
        }
    }

    let client = reqwest::Client::new();

    let (openaec_result, anthropic_result) = tokio::join!(
        fetch_openaec_packages(&client),
        fetch_anthropic_packages(&client),
    );

    let mut packages = Vec::new();

    if let Ok(openaec) = openaec_result {
        packages.extend(openaec);
    }
    if let Ok(anthropic) = anthropic_result {
        packages.extend(anthropic);
    }

    if packages.is_empty() {
        if let Some(cached) = get_cached(u64::MAX) {
            return Ok(cached);
        }
        return Err("No packages found and no cache available".to_string());
    }

    set_cache(&packages);
    Ok(packages)
}

#[tauri::command]
pub fn get_cached_registry() -> Result<Vec<RegistryPackage>, String> {
    get_cached(u64::MAX).ok_or_else(|| "No cached registry data".to_string())
}
